File: src/commands/adopt.ts
import { exists, walk } from "@std/fs";
import { normalize, resolve } from "@std/path";
import { blue, green, red } from "@std/fmt/colors";
import { recreateFile } from "../core/utils.ts";
import type { Logger } from "../core/logger.ts";

type Args = {
  from?: string;
  extra: {
    homedir: string;
    dotfilesFilesDir: string;
    logger?: Logger;
    color?: boolean;
  };
};

export default async (
  { from, extra: { homedir, dotfilesFilesDir, logger, color } }: Args,
) => {
  if (from) {
    from = resolve(normalize(from));
  } else {
    from = dotfilesFilesDir;
  }

  if (!await exists(from, { isReadable: true })) {
    return new Error(
      `Path ${color ? blue(from) : from} does not exists or is not readable`,
    );
  }

  if (from !== dotfilesFilesDir) {
    if (!from.startsWith(homedir)) {
      return new Error(
        `Path ${color ? blue(from) : from} is not a subpath of ${
          color ? blue(homedir) : homedir
        }`,
      );
    }

    if (from.startsWith(dotfilesFilesDir)) {
      return new Error(
        `Path ${color ? blue(from) : from} can not be a subpath of ${
          color ? blue(dotfilesFilesDir) : dotfilesFilesDir
        }`,
      );
    }
  }

  if (await exists(from, { isFile: true })) {
    const to = resolve(normalize(from.replace(homedir, dotfilesFilesDir)));

    await logger?.logNoNl.log(
      `Adopting ${color ? blue(from) : from} to ${color ? blue(to) : to} ...`,
    );

    try {
      await recreateFile(from, to);

      await logger?.log.log(color ? green(" ✓") : " ✓");
    } catch (error) {
      await logger?.log.log(color ? red(" ✕") : " ✕");

      return new Error(`Error adopting ${color ? blue(from) : from}`, {
        cause: error,
      });
    }

    return true;
  }

  const errors: Error[] = [];

  for await (
    const file of walk(from, { includeDirs: false, includeSymlinks: false })
  ) {
    let origin: string;
    let destionation: string;

    if (file.path.startsWith(dotfilesFilesDir)) {
      origin = file.path.replace(dotfilesFilesDir, homedir);
      destionation = file.path;
    } else {
      origin = file.path;
      destionation = file.path.replace(homedir, dotfilesFilesDir);
    }

    await logger?.logNoNl.log(
      `Adopting ${color ? blue(origin) : origin} to ${
        color ? blue(destionation) : destionation
      } ...`,
    );

    try {
      await recreateFile(origin, destionation);

      await logger?.log.log(color ? green(" ✓") : " ✓");
    } catch (error) {
      await logger?.log.log(color ? red(" ✕") : " ✕");

      errors.push(
        new Error(`Error adopting ${color ? blue(origin) : origin}`, {
          cause: error,
        }),
      );
    }
  }

  if (errors.length > 0) {
    return new AggregateError(errors, "Error adopting directory");
  }

  return true;
};

File: src/commands/adopt.go
package commands

import (
	"errors"
	"fmt"
	"io/fs"
	"path/filepath"
	"strings"

	"github.com/fatih/color"
	"github.com/gookit/goutil/fsutil"
	"github.com/m4rc3l05/dots/src/core"
)

type AdoptArgsExtra struct {
	Homedir          string
	DotfilesFilesDir string
}

type AdoptArgs struct {
	From  string
	Extra AdoptArgsExtra
}

func (c Commands) Adopt(args AdoptArgs) (bool, error) {
	fromFormatted, err := filepath.Abs(args.From)
	if err != nil {
		return false, err
	}

	args.From = fromFormatted

	if !fsutil.PathExist(args.From) || !core.IsPathReadable(args.From) {
		return false, fmt.Errorf(
			"path %s does not exists or is not readable",
			color.BlueString(args.From),
		)
	}

	if args.From != args.Extra.DotfilesFilesDir {
		if !strings.HasPrefix(args.From, args.Extra.Homedir) {
			return false, fmt.Errorf(
				"path %s is not a subpath of %s",
				color.BlueString(args.From), color.BlueString(args.Extra.Homedir),
			)
		}

		if strings.HasPrefix(args.From, args.Extra.DotfilesFilesDir) {
			return false, fmt.Errorf(
				"path %s can not be a subpath of %s",
				color.BlueString(args.From), color.BlueString(args.Extra.DotfilesFilesDir),
			)
		}
	}

	if fsutil.IsFile(args.From) {
		to := strings.Replace(args.From, args.Extra.Homedir, args.Extra.DotfilesFilesDir, 1)

		c.Logger.Log(
			"Adopting %s to %s ...",
			color.BlueString(args.From),
			color.BlueString(to),
		)

		if err := core.RecreateFile(args.From, to); err != nil {
			c.Logger.Lognl(color.GreenString(" ✕"))

			return false, errors.Join(fmt.Errorf(
				"error adopting %s to %s",
				color.BlueString(args.From),
				color.BlueString(to),
			), err)
		}

		c.Logger.Lognl(color.GreenString(" ✓"))

		return true, nil
	}

	var errorsArr []error

	err = filepath.WalkDir(args.From, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if !d.Type().IsRegular() {
			return nil
		}

		var origin string
		var destination string

		if strings.HasPrefix(path, args.Extra.DotfilesFilesDir) {
			origin = strings.Replace(path, args.Extra.DotfilesFilesDir, args.Extra.Homedir, 1)
			destination = path
		} else {
			origin = path
			destination = strings.Replace(path, args.Extra.Homedir, args.Extra.DotfilesFilesDir, 1)
		}

		c.Logger.Log(
			"Adopting %s to %s ...",
			color.BlueString(origin),
			color.BlueString(destination),
		)

		if err := core.RecreateFile(origin, destination); err != nil {
			c.Logger.Lognl(color.RedString(" ✕"))

			errorsArr = append(
				errorsArr,
				errors.Join(fmt.Errorf(
					"error adopting %s to %s",
					color.BlueString(origin),
					color.BlueString(destination),
				), err),
			)
		} else {
			c.Logger.Lognl(color.GreenString(" ✓"))
		}

		return nil
	})
	if err != nil {
		errorsArr = append(errorsArr, err)
	}

	if len(errorsArr) > 0 {
		errorsArr = append([]error{errors.New("error adopting directory")}, errorsArr...)
	}

	return len(errorsArr) <= 0, errors.Join(errorsArr...)
}
